clients/roscpp/include/ros/timer_manager.h

```diff
@@ -578,14 +586,14 @@ void TimerManager<T, D, E>::threadFunc()
       // since simulation time may be running faster than real time.
       if (!T::isSystemTime())
       {
-        timers_cond_.timed_wait(lock, boost::posix_time::milliseconds(1));
+        timers_cond_.wait_for(lock, boost::chrono::milliseconds(1));
       }
       else
       {
         // On system time we can simply sleep for the rest of the wait time, since anything else requiring processing will
         // signal the condition variable
-        int64_t remaining_time = std::max<int64_t>((sleep_end - current).toSec() * 1000.0f, 1);
-        timers_cond_.timed_wait(lock, boost::posix_time::milliseconds(remaining_time));
+        typename TimerManagerTraits<T>::time_point end_tp(boost::chrono::nanoseconds(sleep_end.toNSec()));
+        timers_cond_.wait_until(lock, end_tp);
```

Comment:
What motivated the change to use `wait_until(end)` instead of `wait_for(duration)`? Without this change it seems `TimerManagerTraits` wouldn't be necessary either.

Reply:
Yes, that should also work, indeed. I copied this section from [steady_timer.cpp](https://github.com/ros/ros_comm/pull/1878/files#diff-e304e755ded42c610ed4975cd99efd75L106-L109) and introduced the traits to avoid having to specialize the whole method.

Unfortunately I don't have time right now to test the proposed patch. It is important not to use `timed_wait()`, but `wait_for(duration)` as suggested.

Reply:
I implemented that idea in 2a67cf6780c32fef3af96fac9da405f4e6eb0298.